rasa/utils/tf_utils.py

```diff
@@ -0,0 +1,898 @@
+from collections import namedtuple
+import logging
+import numpy as np
+from tqdm import tqdm
+from sklearn.model_selection import train_test_split
+import tensorflow as tf
+from tensor2tensor.models.transformer import (
+    transformer_base,
+    transformer_prepare_encoder,
+    transformer_encoder,
+)
+from tensor2tensor.layers.common_attention import large_compatible_negative
+from rasa.utils.common import is_logging_disabled
+import typing
+from typing import List, Optional, Text, Dict, Tuple, Union, Generator, Callable
+
+if typing.TYPE_CHECKING:
+    from tensor2tensor.utils.hparam import HParams
+
+# avoid warning println on contrib import - remove for tf 2
+tf.contrib._warning = None
+logger = logging.getLogger(__name__)
+
+# namedtuple for all tf session related data
+SessionData = namedtuple("SessionData", ("X", "Y", "labels"))
+
+
+# noinspection PyPep8Naming
+def train_val_split(
+    session_data: "SessionData", evaluate_on_num_examples: int, random_seed: int
+) -> Tuple["SessionData", "SessionData"]:
+    """Create random hold out validation set using stratified split."""
+
+    label_counts = dict(
+        zip(*np.unique(session_data.labels, return_counts=True, axis=0))
+    )
+    counts = np.array([label_counts[label] for label in session_data.labels])
+
+    multi_X = session_data.X[counts > 1]
+    multi_Y = session_data.Y[counts > 1]
+    multi_labels = session_data.labels[counts > 1]
+
+    solo_X = session_data.X[counts == 1]
+    solo_Y = session_data.Y[counts == 1]
+    solo_labels = session_data.labels[counts == 1]
+
+    (X_train, X_val, Y_train, Y_val, labels_train, labels_val) = train_test_split(
+        multi_X,
+        multi_Y,
+        multi_labels,
+        test_size=evaluate_on_num_examples,
+        random_state=random_seed,
+        stratify=multi_labels,
+    )
+    X_train = np.concatenate([X_train, solo_X])
+    Y_train = np.concatenate([Y_train, solo_Y])
+    labels_train = np.concatenate([labels_train, solo_labels])
+
+    return (
+        SessionData(X=X_train, Y=Y_train, labels=labels_train),
+        SessionData(X=X_val, Y=Y_val, labels=labels_val),
+    )
+
+
+def shuffle_session_data(session_data: "SessionData") -> "SessionData":
+    """Shuffle session data."""
+
+    ids = np.random.permutation(len(session_data.X))
+    return SessionData(
+        X=session_data.X[ids], Y=session_data.Y[ids], labels=session_data.labels[ids]
+    )
+
+
+# noinspection PyPep8Naming
+def gen_batch(
+    session_data: "SessionData",
+    batch_size: int,
+    batch_strategy: Text = "sequence",
+    shuffle: bool = False,
+) -> Generator[Tuple["np.ndarray", "np.ndarray"], None, None]:
+    """Generate batches."""
+
+    if shuffle:
+        session_data = shuffle_session_data(session_data)
+
+    if batch_strategy == "balanced":
+        num_examples = len(session_data.X)
+        unique_labels, counts_labels = np.unique(
+            session_data.labels, return_counts=True, axis=0
+        )
```

Comment:
Can we abstract out the reformation of `SessionData` based on stratified sampling in a different function and call that function within this `if` construct?